consensus/consensus.go

```diff
@@ -1,103 +1,237 @@
+	// Check if the size of the transaction list is equal to 1.
+	if len(gb.Transactions) != 1 {
+		return false, BadGenesisTransactions
 	}
 
-	// Check if the block's target is equal to the network's current target
-	if b.Target != CurrentTarget() {
-		log.Error("Invalid Target")
+	// Check if the transaction is a valid cloud base transaction.
+	if valid, code := VerifyCloudBase(bc, gb.Transactions[0]); !valid {
+		log.Errorf("Invalid CloudBase, CloudBaseTransactionCode: %d", code)
+		return false, BadGenesisCloudBaseTransaction
+	}
+
+	// Check that the target is within the min and max difficulty levels.
+	// TODO: CurrentTarget() is used because the difficulty is static for v1 of
+	// cumulus. Once the difficulty is dynamic, the target would need to be
+	// compared to the target calculated for that specific block.
+	target := blockchain.HashToBigInt(gb.Target)
+	if target.Cmp(c.MaxTarget) == 1 ||
+		target.Cmp(c.MinTarget) == -1 ||
+		target.Cmp(blockchain.HashToBigInt(CurrentTarget())) != 0 {
+		return false, BadGenesisTarget
+	}
+
+	// Check that time is not greater than current time or equal to 0.
+	if uint32(gb.Time) == 0 {
+		return false, BadGenesisTime
+	}
+
+	return true, ValidGenesisBlock
+}
+
+// VerifyBlock checks whether a block is valid.
+func VerifyBlock(bc *blockchain.BlockChain,
+	b *blockchain.Block) (bool, BlockCode) {
+
+	// Check if the block is equal to nil.
+	if b == nil {
+		return false, NilBlock
+	}
+
+	// Check if the block is the genesis block.
+	if b.BlockHeader.BlockNumber == 0 || bc.Blocks[0] == b {
+		if valid, code := VerifyGenesisBlock(bc, b); !valid {
+			log.Errorf("Invalid GenesisBlock, GenesisBlockCode: %d", code)
+			return false, BadGenesisBlock
+		}
+		return true, ValidBlock
+	}
+
+	// Check that block number is between 0 and max blocks.
+	ix := b.BlockNumber - 1
+	if int(ix) > len(bc.Blocks)-1 || ix < 0 {
+		return false, BadBlockNumber
+	}
+
+	// Check that block number is one greater than last block
+	lastBlock := bc.Blocks[ix]
+	if lastBlock.BlockNumber != ix {
+		return false, BadBlockNumber
+	}
+
+	// Check that the first transaction is a CloudBase transaction
+	if valid, code := VerifyCloudBase(bc, b.GetCloudBaseTransaction()); !valid {
+		log.Errorf("Invalid CloudBase, CloudBaseTransactionCode: %d", code)
+		return false, BadCloudBaseTransaction
+	}
+
+	// Verify every Transaction in the block.
+	for _, t := range b.Transactions[1:] {
+		if valid, code := VerifyTransaction(bc, t); !valid {
+			log.Errorf("Invalid Transaction, TransactionCode: %d", code)
+			return false, BadTransaction
+		}
+	}
+
+	// Check that the target is within the min and max difficulty levels and
+	// that the target is correct.
+	// TODO: CurrentTarget() is used because the difficulty is static for v1 of
+	// cumulus. Once the difficulty is dynamic, the target would need to be
+	// compared to the target calculated for that specific block.
+	target := blockchain.HashToBigInt(b.Target)
+	if target.Cmp(c.MaxTarget) == 1 ||
```

Comment:
Doesn't this work:

```
if target.Cmp(blockchain.HashToBigInt(CurrentTarget())) != 0 {
    return false, BadTarget
}
```

Why would the `CurrentTarget()` return something out of range?